fix Issue 23058 - importC: cannot take address inside multi-dimension…

WalterBright opened the pull request:
…al array at compile time

ljmf00 committed:
chore(console): make isatty extern signature private

It's also annotated as a @trusted @nogc nothrow interface.

Reference: https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/posix-isatty?view=msvc-170
Signed-off-by: Luís Ferreira <contact@lsferreira.net>

WalterBright committed:
add bitfields to D (#13568)

ibuclaw committed:
test: Add RUNNABLE_PHOBOS_TEST and COMPILABLE_MATH_TEST comments (#13012)

This highlights the remaining tests that still have phobos dependencies

Yoplitein committed:
Amend prior fix for issue 22877

Aligns ImportC's treatment of string literals with C11 by typing them as static arrays
Also fixes `typeof(&"")` returning string length instead of pointer size

ibuclaw committed:
Move driver and codegen fields from Param to DMDparam (#13796)

* Move driver and codegen fields from Param to DMDparam

* Rename other additions from dmdParams to driverParams

* Fix conflict

* Fix conflict

Co-authored-by: RazvanN7 <razvan.nitu1305@gmail.com>

Yoplitein committed:
Expand fix for issue 22070 to also allow array literals as lvalues

Check for and restrict nested address-of expressions on literals, such as `&(&"")`

thewilsonator committed:
Reduce Indentation in `dmd/dsymbolsem.d`

thewilsonator committed:
reduce indentation in `dmd/dsymbolsem.d

thewilsonator committed:
reduce indentation in `dmd/dsymbolsem.d`

thewilsonator committed:
reduce indenting in `dmd/dsymbolsem.d`

thewilsonator committed:
reduce indenting in `dmd/dsymbolsem.d` (#14034)

dkorpel committed:
Fix issue 23068 - BetterC does not respect -checkaction=halt

thewilsonator committed:
reduce indenting in `dmd/dsymbolsem.d`

WalterBright committed:
fix Issue 23050 - Incorrect disassembly of code with -vasm and 0xBE and 0xBF opcodes (#14037)

thewilsonator committed:
reduce indenting in `tocvdebug.d` (#14039)

* reduce indenting in `tocvdebug.d`

* Missing brace

* remove unreachable

thewilsonator committed:
Consolidate output parameters in global.params (#12617)

dkorpel committed:
Remove visitor from escapeByValue (#14040)

dkorpel committed:
Remove visitor from escapeByRef (#14041)

ibuclaw committed:
refactor: Replace toParentXXX.isAggregateDeclaration with isMemberXXX

WalterBright committed:
fix Issue 23058 - importC: cannot take address inside multi-dimensional array at compile time

WalterBright commented:
Test suite failures are networking problems with Azure.

thewilsonator committed:
Make `global.params.mixinOut` an `Output` (#14048)

thewilsonator committed:
Compile `tocvdebug.d` for all platforms

Make DMD's anemic cross-compilation, slightly less anemic.

thewilsonator committed:
remove the markdown related switches

This has been the default for a couple of years now.

dkorpel committed:
Move scope parameter inference to escape.d

kinke committed:
d_do_test: Respect CppRuntime_Clang targets by linking libc++, not libstdc++

WalterBright committed:
ImportC: run cpp on Posix .c files (#13955)

dkorpel commented on src/dmd/optimize.d:
```diff
@@ -558,6 +558,41 @@ Expression Expression_optimize(Expression e, int result, bool keepLvalue)
+
+                    if (index < 0 || index > dim)
+                    {
+                        e.error("array index %lld is out of bounds `[0..%lld]`", index, dim);
```
Not covered by tests

dkorpel commented on src/dmd/optimize.d:
```diff
@@ -558,6 +558,41 @@ Expression Expression_optimize(Expression e, int result, bool keepLvalue)
+                    const offset = mulu(index, ts.nextOf().size(e.loc), overflow); // offset = index*size
+                    if (overflow)
+                    {
+                        e.error("array offset overflow");
```
Not covered by tests

WalterBright replied:
It's difficult to come up with a test for that.

WalterBright replied:
In many cases, I have been stymied by trying to trip these, as the overflows happen before it gets here. But the check should still be there, for defensive programming purposes.

dkorpel commented on test/fail_compilation/fail13902.d:
```diff
@@ -57,7 +57,7 @@ fail_compilation/fail13902.d(88): Error: Using the result of a comma expression
 fail_compilation/fail13902.d(75): Error: returning `& x` escapes a reference to parameter `x`
 fail_compilation/fail13902.d(76): Error: returning `&s1.v` escapes a reference to parameter `s1`
 fail_compilation/fail13902.d(81): Error: returning `& sa1` escapes a reference to parameter `sa1`
-fail_compilation/fail13902.d(82): Error: returning `&sa2[0][0]` escapes a reference to parameter `sa2`
```
This PR introduces a regression in D error messages.
```D
int* f()
{
    int[2][2] sa2;
    return &sa2[1][1];
}
```

> returning `(& sa2+12)` escapes a reference to local variable `sa2`

That's not what's being returned.

WalterBright replied:
yes, it is what is being returned.

dkorpel commented on src/dmd/optimize.d:
```diff
@@ -558,6 +558,41 @@ Expression Expression_optimize(Expression e, int result, bool keepLvalue)
         {
+            if (ae.e2.isIntegerExp() && ae.e1.isIndexExp())
+            {
+                /* Rewrite `(a[i])[index]` to `(&a[i]) + index*size`
```
The comment should say why this is done

WalterBright replied:
The reason is the same as for every other rewrite - to put it in simpler terms that the rest of the code can handle. I.e. "lowering".

ibuclaw committed:
test/compilable/test17434a.d: Add missing EXTRA_FILES directive

ljmf00 committed:
Fix issue 22911: dtoh: make include directives sorted for generated headers (#13867)

Signed-off-by: Luís Ferreira <contact@lsferreira.net>

dkorpel committed:
Issue 17374 - Begin improving inferred attribute error message for safety (#13957)

Yoplitein committed:
ImportC: recognize typeof operator (#14000)

RazvanN7 commented:
ping @WalterBright

ljmf00 committed:
iasmdmd: replace init_optab() with internal checks (#13822)

init_optab() doesn't need to run everytime the compiler initiates, only
internally on compilation/testing.

Signed-off-by: Luís Ferreira <contact@lsferreira.net>

dkorpel committed:
Put error messages inside setUnsafe()

dkorpel committed:
Put error messages inside setUnsafe() (escape.d)

dkorpel committed:
Fix issue 23042 - -betterC still includes RTInfo

ibuclaw committed:
fail_compilation/test23037.c: Adjust test so it fails consistently on all platforms

kinke committed:
Don't require extern(C) for CRT con/destructors

The default signature is `void ()`, which e.g. works with C++ and D
linkages just fine too. IIRC, an MSVCRT constructor is called with
the main() arguments, so only require `extern(C)` for signatures
with parameters - and newly check that all functions return void.

extern(D) CRT con/destructors are nice because they don't pollute
the global C namespace and prevent symbol conflicts at link-time.

WalterBright committed:
ImportC: add Win64 preprocessor support (#14059)

dkorpel committed:
Put error messages in setUnsafe() 2

dkorpel committed:
Put error messages in setUnsafeDip1000() (#14087)

dkorpel committed:
Use setUnsafeDIP1000 in escape.d more consistently

ibuclaw committed:
dmd.cpreprocess: Don't delete non-generated files (#14061)

rainers committed:
C preprocessor: cl.exe needs to be run from the path appropriate for the target architecture

allow overriding the cpp command line with CPPCMD

WalterBright commented:
gawd, not again.

WalterBright commented:
rebasing it back to stable made an even worse mess.

WalterBright commented:
Rebooted as https://github.com/dlang/dmd/pull/14094
